File: frontend/src/types/commerce/promocode.ts
import {IUser} from "types/core/user";
import {ICurrency, IProduct} from "types/commerce/shop";

export interface IPromocode {
    id: number
    name: string;
    code: string;
    description?: string;
    discounts: IPromocodeDiscount[];
    discount_type: 'percentage' | 'fixed_amount';
    start_date?: string; // дата и время
    end_date?: string; // дата и время
}


export interface IPromocodeDiscount {
    id: number;
    product: IProduct | number;
    currency: ICurrency;
    amount: number;
    max_usage?: number;
    max_usage_per_user?: number;
    interval_days?: number;
    specific_users?: IUser[];
}

export interface IPromocodeUsage {
    id: number
    promocode: IPromocode;
    user: IUser;
    created_at: string;
}

File: backend/apps/commerce/admin/promocode.py
from django.contrib import admin
from import_export.admin import ImportExportModelAdmin

from apps.commerce.models.promocode import (
    Promocode, PromocodeUsage, PromocodeProductDiscount
)


@admin.register(Promocode)
class PromocodeAdmin(ImportExportModelAdmin):
    list_display = (
        'name',
        'code',
        'discount_type',
        'start_date',
        'end_date'
    )
    search_fields = ('name', 'code')
    list_filter = ('discount_type', 'start_date', 'end_date')
    ordering = ('-start_date',)
    fieldsets = (
        (None, {
            'fields': (
                'name',
                'code',
                'description',
                'discounts',
                'discount_type',
                'start_date',
                'end_date'
            )
        }),
    )


@admin.register(PromocodeProductDiscount)
class PromocodeProductDiscountAdmin(admin.ModelAdmin):
    list_display = (
        'product',
        'max_usage',
        'interval_days',
        'max_usage_per_user',
        'amount', 'currency',
    )
    search_fields = (
        'products__name',
    )
    list_filter = (
        'currency',
        'max_usage',
        'interval_days'
    )
    fieldsets = (
        (None, {
            'fields': (
                'product',
                'amount', 'currency',
                'max_usage',
                'interval_days',
                'max_usage_per_user',
                'specific_users',
            )
        }),
    )
    filter_horizontal = ('specific_users',)


@admin.register(PromocodeUsage)
class PromocodeUsageAdmin(admin.ModelAdmin):
    list_display = (
        'promocode',
        'user',
        'created_at'
    )
    search_fields = (
        'promocode__code',
        'user__username',
    )
    list_filter = ('created_at',)
    ordering = ('-created_at',)
